signer/single_signer_validator.go

```diff
@@ -1,42 +1,116 @@
 package signer
 
 import (
+	"fmt"
+	"os"
 	"sync"
 
 	tmcrypto "github.com/tendermint/tendermint/crypto"
+	tmjson "github.com/tendermint/tendermint/libs/json"
+	tmprivval "github.com/tendermint/tendermint/privval"
 	tmproto "github.com/tendermint/tendermint/proto/tendermint/types"
-	tm "github.com/tendermint/tendermint/types"
 )
 
 var _ PrivValidator = &SingleSignerValidator{}
 
 // SingleSignerValidator guards access to an underlying PrivValidator by using mutexes
 // for each of the PrivValidator interface functions
 type SingleSignerValidator struct {
-	PrivValidator tm.PrivValidator
-	pvMutex       sync.Mutex
+	config     *RuntimeConfig
+	chainState sync.Map
+	pubKey     tmcrypto.PubKey
+}
+
+type SingleSignerChainState struct {
+	filePV  *tmprivval.FilePV
+	pvMutex sync.Mutex
+}
+
+// NewSingleSignerValidator constructs a validator for single-sign mode (not recommended).
+// NewThresholdValidator is recommended, but single-sign mode can be used for convenience.
+func NewSingleSignerValidator(config *RuntimeConfig) (*SingleSignerValidator, error) {
+	pv := &SingleSignerValidator{
+		config: config,
+	}
+
+	if err := pv.loadPubKey(); err != nil {
+		return nil, fmt.Errorf("failed to load priv validator key: %w", err)
+	}
+
+	return pv, nil
+}
+
+func (pv *SingleSignerValidator) loadPubKey() error {
+	keyFile := pv.config.KeyFilePathSingleSigner()
+
+	keyJSONBytes, err := os.ReadFile(keyFile)
+	if err != nil {
+		return err
+	}
+	pvKey := tmprivval.FilePVKey{}
+	err = tmjson.Unmarshal(keyJSONBytes, &pvKey)
+	if err != nil {
+		return err
+	}
+
+	pv.pubKey = pvKey.PrivKey.PubKey()
+
+	return nil
 }
 
 // GetPubKey implements types.PrivValidator
 func (pv *SingleSignerValidator) GetPubKey() (tmcrypto.PubKey, error) {
-	pv.pvMutex.Lock()
-	defer pv.pvMutex.Unlock()
-	return pv.PrivValidator.GetPubKey()
+	return pv.pubKey, nil
 }
 
 // SignVote implements types.PrivValidator
 func (pv *SingleSignerValidator) SignVote(chainID string, vote *tmproto.Vote) error {
-	pv.pvMutex.Lock()
-	defer pv.pvMutex.Unlock()
-	return pv.PrivValidator.SignVote(chainID, vote)
+	chainState, err := pv.loadChainStateIfNecessary(chainID)
+	if err != nil {
+		return err
+	}
+	chainState.pvMutex.Lock()
+	defer chainState.pvMutex.Unlock()
+	return chainState.filePV.SignVote(chainID, vote)
 }
```

Comment:
I swear I asked this already but cannot find the GH comment anywhere. Apologies about the double ask.

What does this mutex protect? I see it's locking per chainID, which is fine.

If it's to protect against a double sign, I don't see how it would work. That's the caller's responsibility. 

The caller could do:

```go
vote := getVoteSomehow()
for {
   validator.SignVote("cosmoshub-4", vote)
}
```

The same applies for `SignProposal`.

Reply:
Thanks for that explanation. That last comment would be great in go doc comments.